### PR TITLE
random helpers

### PR DESCRIPTION
**Pull request by arnetheduck**
* arrayops as a home for small array/openArray utilities
* assign2 - a replacement for genericAssign and assignment operators in
general which in nim are very slow
* use assign2 in a few places to speed things up

### CODE REVIEW

File: stew/assign2.nim [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    for t, s in fields(tgt, src):
+      when supportsCopyMem(type s) and sizeof(s) <= sizeof(int) * 2:
+        t = s # Shortcut
+      else:
+        assign(t, s)
+  elif T is seq:
+    assign(tgt, src.toOpenArray(0, src.high))
+  elif T is ref:
+    tgt = src
+  elif compiles(distincBase(T)):
```
Comment (zah) [inline stew/assign2.nim:47]: This goes a little bit against the spirit of distinct types which are sometimes introduced specifically to provide different behavior for certain operations (in this case, assignment).

**Review by zah — APPROVED** [review]

File: stew/assign2.nim (hunk 1 above)
Comment (arnetheduck) [reply stew/assign2.nim:47]: `=` works for `distinct` types out-of-the-box - `assign` is pretty a drop-in replacement that fixes the massive inefficiency in the std lib - unfortunately we can't overload `=` and the c-generated assignments to use it - it really needs an upstream fix at which point this function/module becomes obsolete (https://github.com/nim-lang/Nim/issues/15571)

Comment (arnetheduck) [reply stew/assign2.nim:47]: ah, you mean that we should try to detect specific `=` overloads for those distinct types? hm.. possible?

Comment (arnetheduck) [reply stew/assign2.nim:47]: if that's the case, should distinct types with `=` really `supportsCopyMem`?

File: stew/assign2.nim [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  std/typetraits,
+  ./shims/macros
+
+{.push raises: [Defect].}
+
+func assignImpl[T](tgt: var openArray[T], src: openArray[T]) =
+  mixin assign
+  when supportsCopyMem(T):
+    if tgt.len > 0:
+      copyMem(addr tgt[0], unsafeAddr src[0], sizeof(tgt[0]) * tgt.len)
```
Comment (ringabout) [inline stew/assign2.nim:11]: Is it possible that the memory of `tgt` and `src` overlaps? ref https://github.com/nim-lang/Nim/pull/21661

Comment (arnetheduck) [reply stew/assign2.nim:11]: that's a good point, I didn't consider overlapping arrays (including self-assignment)

Comment (arnetheduck) [reply stew/assign2.nim:11]: https://github.com/status-im/nim-stew/pull/182

**Pull request merged**